cirq/ops/diagonal_gate.py

```diff
@@ -0,0 +1,196 @@
+            return NotImplemented
+        angles = []
+        for angle in self._diag_angles_radians:
+            mul_angle = protocols.mul(angle, exponent, NotImplemented)
+            angles.append(mul_angle)
+        return DiagonalGate(angles)
+
+    def _value_equality_values_(self) -> Any:
+        return tuple(self._diag_angles_radians)
+
+    def _decompose_for_basis(
+        self, index: int, bit_flip: int, theta: float, qubits: Sequence['cirq.Qid']
+    ) -> Iterator[Union['cirq.ZPowGate', 'cirq.CXPowGate']]:
+        if index == 0:
+            return []
+        largest_digit = self._num_qubits_() - (len(bin(index)) - 2)
+        yield common_gates.rz(2 * theta)(qubits[largest_digit])
+        _flip_bit = self._num_qubits_() - bit_flip - 1
+        if _flip_bit < largest_digit:
+            yield common_gates.CNOT(qubits[largest_digit], qubits[_flip_bit])
+        elif _flip_bit > largest_digit:
+            yield common_gates.CNOT(qubits[_flip_bit], qubits[largest_digit])
+
+    def _decompose_(self, qubits: Sequence['cirq.Qid']) -> 'cirq.OP_TREE':
+        """Decompose the n-qubit diagonal gates into CNOT and Rz gates.
+
+        A 3 qubits decomposition looks like
+        0: ───────────────────────────────────X───Rz(6)───X───Rz(7)───X───Rz(5)───X───Rz(4)───
+                                              │           │           │           │
+        1: ───────────X───Rz(3)───X───Rz(2)───@───────────┼───────────@───────────┼───────────
+                      │           │                       │                       │
+        2: ───Rz(1)───@───────────@───────────────────────@───────────────────────@───────────
+
+        where the angles in Rz gates are corresponding to the fast-walsh-Hadamard transfrom
+        of diagonal_angles in the Gray Code order.
+
+        For n qubits decomposition looks similar but with 2^n-1 Rz gates and 2^n-2 CNOT gates.
+
+        The algorithm is implemented according to the paper:
+            Welch, Jonathan, et al. "Efficient quantum circuits for diagonal unitaries without
+            ancillas." New Journal of Physics 16.3 (2014): 033040.
+            https://iopscience.iop.org/article/10.1088/1367-2630/16/3/033040/meta
+        """
+        if protocols.is_parameterized(self):
+            return NotImplemented
+
+        n = self._num_qubits_()
+        hat_angles = _fast_walsh_hadamard_transform(self._diag_angles_radians) / (2 ** n)
+
+        # There is one global phase shift between unitary matrix of the diagonal gate and the
+        # decomposed gates. On its own it is not physically observable. However, if using this
+        # diagonal gate for sub-system like controlled gate, it is no longer equivalent. Hence,
+        # we add global phase.
+        decomposed_circ = [global_phase_op.GlobalPhaseOperation(np.exp(1j * hat_angles[0]))]
+        for i, bit_flip in _gen_gray_code(n):
+            decomposed_circ.extend(self._decompose_for_basis(i, bit_flip, -hat_angles[i], qubits))
```

Comment:
```suggestion
            decomposed_circ.append(self._decompose_for_basis(i, bit_flip, -hat_angles[i], qubits))
```

Reply:
oh, or `+=`

Reply:
Actually, question: this surely works but why prefer yy += list(xx) than yy.extend(xx)?

Reply:
`extend(list(xx))` would work - `+=` is shorter...but without the `list()` it is failing - CI checks are red because of that.

Reply:
I think it is the invariance list assumption in mypy. I think denote the decomposed_circ: List[Any] will solve the problem.